### PR TITLE
[SPARK-20442][PYTHON][DOCS] Fill up documentations for functions in Column API in PySpark

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to fill up the documentation with examples for `bitwiseOR`, `bitwiseAND`, `bitwiseXOR`. `contains`, `asc` and `desc` in `Column` API.

Also, this PR fixes minor typos in the documentation and matches some of the contents between Scala doc and Python doc.

Lastly, this PR suggests to use `spark` rather than `sc` in doc tests in `Column` for Python documentation.

## How was this patch tested?

Doc tests were added and manually tested with the commands below:

`./python/run-tests.py --module pyspark-sql`
`./python/run-tests.py --module pyspark-sql --python-executable python3`
`./dev/lint-python`

Output was checked via `make html` under `./python/docs`. The snapshots will be left on the codes with comments.

### CODE REVIEW

File: python/pyspark/sql/column.py [hunk 1]
```diff
@@ -527,7 +584,7 @@ def _test():
         .appName("sql.column tests")\
         .getOrCreate()
     sc = spark.sparkContext
-    globs['sc'] = sc
+    globs['spark'] = spark
```
Comment (HyukjinKwon) [inline python/pyspark/sql/column.py]: I removed `sc` and replaced it to `spark` as I think we promote this way up to my knowledge.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitwiseExpressions.scala [hunk 2]
```diff
@@ -86,7 +86,7 @@ case class BitwiseOr(left: Expression, right: Expression) extends BinaryArithmet
 }
 
 /**
- * A function that calculates bitwise xor of two numbers.
+ * A function that calculates bitwise xor({@literal ^}) of two numbers.
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/bitwiseExpressions.scala]: Matching it up with `BitwiseAnd` and `BitwiseOr` where

> A function that calculates bitwise and(&) of two numbers.

> A function that calculates bitwise or(|) of two numbers.

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala [hunk 3]
```diff
@@ -1008,7 +1009,7 @@ class Column(val expr: Expression) extends Logging {
   def cast(to: String): Column = cast(CatalystSqlParser.parseDataType(to))
 
   /**
-   * Returns an ordering used in sorting.
+   * Returns a sort expression based on the descending order of the column.
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: This and the similar instances below are matched with `functions.scala`. They look calling the same ones.

> Returns a sort expression based on the descending order of the column.

> Returns a sort expression based on the descending order of the column,
> and null values appear before non-null values.

> Returns a sort expression based on the descending order of the column,
> and null values appear after non-null values.

**Review by HyukjinKwon — COMMENTED** [review]
I left some images and comment to make review easier.

File: python/pyspark/sql/column.py [hunk 4]
```diff
@@ -185,17 +185,52 @@ def __contains__(self, item):
...
+    Compute bitwise OR of this expression with another expression.
+
+    :param other: a value or :class:`Column` to calculate bitwise or(|) against
+                  this :class:`Column`.
+
+    >>> from pyspark.sql import Row
+    >>> df3 = spark.createDataFrame([Row(a=170, b=75)])
+    >>> df3.select(df3.a.bitwiseOR(df3.b)).collect()
+    [Row((a | b)=235)]
+    """
```
Comment (HyukjinKwon) [inline python/pyspark/sql/column.py]: This is matched with Scala one. 

> Compute bitwise OR of this expression with another expression

File: python/pyspark/sql/column.py [hunk 5]
```diff
@@ -185,17 +185,52 @@ def __contains__(self, item):
...
+    Compute bitwise AND of this expression with another expression.
+
+    :param other: a value or :class:`Column` to calculate bitwise and(&) against
+                  this :class:`Column`.
+
+    >>> from pyspark.sql import Row
+    >>> df3 = spark.createDataFrame([Row(a=170, b=75)])
+    >>> df3.select(df3.a.bitwiseAND(df3.b)).collect()
+    [Row((a & b)=10)]
+    """
```
Comment (HyukjinKwon) [inline python/pyspark/sql/column.py]: This is matched with Scala one.

> Compute bitwise AND of this expression with another expression

File: python/pyspark/sql/column.py [hunk 6]
```diff
@@ -185,17 +185,52 @@ def __contains__(self, item):
...
+    Compute bitwise XOR of this expression with another expression.
+
+    :param other: a value or :class:`Column` to calculate bitwise xor(^) against
+                  this :class:`Column`.
+
+    >>> from pyspark.sql import Row
+    >>> df3 = spark.createDataFrame([Row(a=170, b=75)])
+    >>> df3.select(df3.a.bitwiseXOR(df3.b)).collect()
+    [Row((a ^ b)=225)]
+    """
```
Comment (HyukjinKwon) [inline python/pyspark/sql/column.py]: This is matched with Scala one.

> Compute bitwise XOR of this expression with another expression

File: python/pyspark/sql/column.py [hunk 7]
```diff
@@ -337,26 +381,39 @@ def isin(self, *cols):
...
+    >>> df2.select(df2.name).orderBy(df2.name.desc()).collect()
+    [Row(name=u'Tom'), Row(name=u'Alice')]
+    """
+
+    asc = ignore_unicode_prefix(_unary_op("asc", _asc_doc))
+    desc = ignore_unicode_prefix(_unary_op("desc", _desc_doc))
 
     _isNull_doc = """
-    True if the current expression is null. Often combined with
-    :func:`DataFrame.filter` to select rows with null values.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/column.py]: `Often combined with :func:`DataFrame.filter` to select rows with null values.` was removed because it looks applying to many other APIs and look too much. It just follows Scala one now.

**Comment by HyukjinKwon** [comment]
cc @srowen, @holdenk, @felixcheung, @map222 and @zero323 who were in related PRs.

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala (hunk 3 above)
Comment (map222) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: Do you want to include example usages, as in the python documentation? E.g. for rlike,
```
   * {{{
   *   // find names that start with "Al"
   *   scala> df.filter( $"name".like("Al%") ).collect()
   *   Array([Alice,1])
   * }}}
```

It would also be helpful for the `startsWith` and `endsWith` functions, where there are two versions, e.g. `startsWith(other: Column)` and `startsWith(literal: String)`.

File: python/pyspark/sql/column.py [hunk 8]
```diff
@@ -288,8 +324,16 @@ def __iter__(self):
...
     []
     """
+    _contains_doc = """
+    Contains the other element. Returns a boolean :class:`Column` based on a string match.
+
+    :param other: string in line
+
+    >>> df.filter(df.name.contains('o')).collect()
+    [Row(age=5, name=u'Bob')]
+    """
```
Comment (map222) [inline python/pyspark/sql/column.py]: Should the `_contains_doc` be before the other docs, to match the order of the function declarations?

**Review by felixcheung — APPROVED** [review]
LGTM @holdenk

File: python/pyspark/sql/column.py [hunk 9]
```diff
@@ -251,15 +285,16 @@ def __iter__(self):
 
     # string methods
     _rlike_doc = """
-    Return a Boolean :class:`Column` based on a regex match.
+    SQL RLIKE expression (LIKE with Regex). Returns a boolean :class:`Column` based on a regex
```
Comment (zero323) [inline python/pyspark/sql/column.py]: Could you clarify that `rlike` uses not Python, but Java regular expressions?

**Review by zero323 — APPROVED** [review]

File: python/pyspark/sql/column.py (hunk 9 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/column.py]: Let's leave so that it indicates the regular expression is in SQL syntax. I would like to keep them identically in most cases to reduce the overhead when someone needs to sweep the documentation.

It looks there are few places that needs the clarification. If this is something that has to be done, then, let's do this in another PR.

File: sql/core/src/main/scala/org/apache/spark/sql/Column.scala (hunk 3 above)
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Column.scala]: Yea, that sounds good in a way but the downside of adding examples is to maintain and keep them up to date. Let's leave them out here as this PR targets to fix Python documentation.

File: python/pyspark/sql/column.py (hunk 8 above)
Comment (HyukjinKwon) [inline python/pyspark/sql/column.py]: Sure.

**Comment by HyukjinKwon** [comment]
Thank you for your review and approval @felixcheung, @zero323 and @map222.

File: python/pyspark/sql/column.py [hunk 10]
```diff
@@ -527,7 +583,7 @@ def _test():
         .appName("sql.column tests")\
         .getOrCreate()
     sc = spark.sparkContext
-    globs['sc'] = sc
+    globs['spark'] = spark
     globs['df'] = sc.parallelize([(2, 'Alice'), (5, 'Bob')]) \
```
Comment (map222) [inline python/pyspark/sql/column.py]: Do you want to update the `globs['df']` definition to `spark.createDataFrame`?

Comment (HyukjinKwon) [inline python/pyspark/sql/column.py]: Maybe we could. I think this is not related with Python documentation fix BTW.

**Comment by HyukjinKwon** [comment]
The point here is to fix Python documentation and it matches if there are some mismatches at `bitwiseOR`, `bitwiseAND`, `bitwiseXOR` `contains`, `asc` and `desc` among `functions.py`, `column.py`, `functions.scala` and `Column.scala`. 

I hope other extra changes do not hold off this PR.

**Comment by HyukjinKwon** [comment]
retest this please

File: python/pyspark/sql/column.py (hunk 9 above)
Comment (zero323) [inline python/pyspark/sql/column.py]: Problem is that in Scala or Java users get regular expressions dialect they expect. In Python they don't. 

But fair enough. Let's leave it for another time.

Comment (HyukjinKwon) [inline python/pyspark/sql/column.py]: Thank you.

**Comment by felixcheung** [comment]
@holdenk do you have bandwidth to review this or ok with me pushing this to master?

**Comment by HyukjinKwon** [comment]
gentle ping @holdenk

File: python/pyspark/sql/column.py [hunk 11]
```diff
@@ -185,17 +185,51 @@ def __contains__(self, item):
...
-    bitwiseAND = _bin_op("bitwiseAND")
-    bitwiseXOR = _bin_op("bitwiseXOR")
+    _bitwiseOR_doc = """
+    Compute bitwise OR of this expression with another expression.
+
+    :param other: a value or :class:`Column` to calculate bitwise or(|) against
+                  this :class:`Column`.
+
+    >>> from pyspark.sql import Row
+    >>> df3 = spark.createDataFrame([Row(a=170, b=75)])
```
Comment (gatorsmile) [inline python/pyspark/sql/column.py]: Why `df3` instead of `df`?

Comment (HyukjinKwon) [inline python/pyspark/sql/column.py]: I think there is a global `df` variable when running doctests and I guess it was avoided to shadow the same name from the outer scope in some doctests whereas other doctests just use it. I get your point. In documentation, we will only see only the code block and I guess using `df` might be slightly better.

AFAIK, usually, Python documentation have self-contained doctests so I don't know which case is better and correct. If you could confirm this, I can sweep it.

Comment (HyukjinKwon) [inline python/pyspark/sql/column.py]: Let me correct this as probably I guess you prefer `df` and I don't have preference.

**Comment by HyukjinKwon** [comment]
@holdenk, @felixcheung and @gatorsmile, could this get merged?

**Comment by holdenk** [comment]
I can take a review look today, sorry this week has been so busy.

**Comment by holdenk** [comment]
Thank you for improving the documentation @HyukjinKwon looks good to me.

File: python/pyspark/sql/column.py [hunk 12]
```diff
@@ -269,17 +305,17 @@ def __iter__(self):
...
+    :param other: string at start of line (do not use a regex `^`)
 
     >>> df.filter(df.name.startswith('Al')).collect()
     [Row(age=2, name=u'Alice')]
     >>> df.filter(df.name.startswith('^Al')).collect()
     []
     """
     _endswith_doc = """
-    Return a Boolean :class:`Column` based on matching end of string.
+    String ends with. Returns a boolean :class:`Column` based on a string match.
```
Comment (holdenk) [inline python/pyspark/sql/column.py]: This seems to be form the function above, but its correct in the code so no worries

**Comment by holdenk** [comment]
And thanks to everyone for reviewing, I'll merge this to master :)
